Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -472,7 +472,114 @@ public struct HTTPHeaders: CustomStringConvertible, ExpressibleByDictionaryLiter
             return result.map { $0[...] }
         }
 
-        return result.flatMap { $0.split(separator: ",").map { $0.trimWhitespace() } }
+        var values: [Substring] = []
+        for result in result {
+            var parser = HTTPHeaderValueParser(result)
+            while let value = parser.nextValue() {
+                values.append(value)
+            }
+        }
+        return values
+    }
+}
+
+private struct HTTPHeaderValueParser {
+    var current: Substring
+
+    init(_ string: String) {
+        self.current = .init(string)
+    }
+
+    mutating func nextValue() -> Substring? {
+        self.popWhitespace()
+        guard !self.current.isEmpty else {
+            return nil
+        }
+        let value: Substring
+        if self.current.first == "\"" {
+            self.pop()
+            guard let nextDoubleQuote = self.nextDoubleQuote() else {
+                return nil
+            }
+            value = self.pop(to: nextDoubleQuote).unescapingQuotes()
+            self.pop()
+            if let comma = self.nextComma() {
+                _ = self.pop(to: comma)
+                self.pop()
+            }
+        } else if let comma = self.nextComma() {
+            value = self.pop(to: comma)
+            self.pop()
+        } else if let whitespace = self.nextWhitespace() {
+            value = self.pop(to: whitespace)
+        } else {
+            value = self.pop(to: self.current.endIndex)
+        }
+        return value
+    }
+
+    private func nextDoubleQuote() -> Substring.Index? {
+        return self.nextDoubleQuote(from: self.current.startIndex)
+    }
+
+    private func nextDoubleQuote(from startIndex: Substring.Index) -> Substring.Index? {
+        guard let nextQuote = self.current[startIndex...].firstIndex(of: "\"") else {
+            return nil
+        }
+        if self.current[self.current.index(before: nextQuote)] == "\\" {
```

Comment:
This is not guaranteed to be safe: if the double-quote is literally at the front of the current string then this will perform an out-of-bounds access. Can you add a test-case that includes parsing the quoted string `""`? It should fail here, and we can then fix it.